### DISCUSSION TITLE
Bertopic semi-supervised text classifier

### DISCUSSION THREAD

**Discussion by matrix00100** [MaartenGr/BERTopic#623] [Q&A]
Hi everyone,

I have table of text data from customer.  I would like to use Bertopic to figure out the topics from the text data and create labels and then train the Bertopic  model with these labels.  After that, I want to use the model to classify new customer messages and count the message topic label.  Is it possible with Bertopic?  If you have any other suggestions how to tackle this, I really appreciate this.

**Comment by MaartenGr** [comment]
> After that, I want to use the model to classify new customer messages and count the message topic label. Is it possible with Bertopic? 

I believe your use case should be possible with BERTopic but I do get the feeling you might have approached it a bit more complex than necessary. If you want BERTopic to figure out topics and create labels, then you can simply use `BERTopic.fit()`. Then, when you want to use that model to classify new customer messages, you can use `BERTopic.transform()`. From your description, I do not think it is necessary to first use `BERTopic.fit()` to figure out the topics and create labels, and then use those labels for another pass of `BERTopic.fit()`. The end result will likely be the same.

**Reply by matrix00100** [reply]
Hello,

Thank you so much for helping me.  I am new to this area.

My intention is using Bertopic to run semi-supervised learning using the weak labels that I create. After that, I will label a few customer messages using those weak labels and do a transfer learning to improve the model accuracy.  Is it possible to do that with Bertopic?  Does Bertopic have accuracy metric?  Again really appreciate your guidance.

**Reply by matrix00100** [reply]
I tried .fit(customer_messages, y=customized_labels).  It gave a error "Length of x = 400000, length of y = 33, while it must be equal".

**Reply by drob-xx** [reply]
So I'm not sure exactly what is going on here, but my guess is that you have a classification rubric that has 33 different classes and 400K documents that you want to place those into? I'm not that familiar with the way BERTopic does guided classification - but you would need to label _each_ of the 400K documents with one your 33 classifications for this to work - in other words the dimension of these two values has to match in n. 

It might be helpful to clearly describe your data and what your business case is to give more context.

**Reply by matrix00100** [reply]
![image](https://user-images.githubusercontent.com/52605129/180235551-404c7a3c-6c84-42d5-a2fa-80da4bebf878.jpeg)

Sorry due to the privacy, I can't post the data.  However, the data set looks similar to the IMDB dataset except no labels(positive or negative).

I want to use label these customer messages with customized labels as  weak labels supervised learning then do transfer learning later with labeled messages when we have the labeled data.

**Reply by drob-xx** [reply]
If you are trying to get binary positive/negative labels for the reviews then as far as I know BERTopic isn't the right tool (but @MaartenGr can be more definitive). 

If this is what your goal is the reason that BERTopic isn't a good choice is that topic models provide word vocabularies affiliated with an underlying topic representation of text. These won't tell you anything about the _kind_ of speech (positive, negative) used, just a summary of _what_ speech is used. There are a ton of articles about sentiment analysis (determining the positive/negative valence of a collection of words) - here's one article <https://medium.com/@pyashpq56/sentiment-analysis-on-imdb-movie-review-d004f3e470bd> that seems decent enough - but it is easy to find more.

If you aren't looking to do sentiment analysis then please let me know what I got wrong and why you think BERTopic is a good choice for your project.

**Reply by matrix00100** [reply]
Oh no I am not looking for sentiment analysis.  I use it as an example for a text column of customer data.

I want to create a model to classify these messages with different labels such a ["sale", "product questions, "return questions",].  The list of labels can be like 30 labels.  

After that, I want to apply this model toward new customer messages  to classify them and count the number of topics for business intelligence.

Hope I make it clearer.  I wish I could share the data but that is against company policy.

**Comment by drob-xx** [comment]
No problem. So when you got the error above:

>I tried .fit(customer_messages, y=customized_labels). It gave a error "Length of x = 400000, length of y = 33, while it must be equal".

What was the content of `customized_labels`, if you want BERTopic to consider your labels when it builds the topics then you would need to supply a label for each document.
